elyra/contents/file_parser.py

```diff
@@ -0,0 +1,279 @@
+#
+# Copyright 2018-2021 Elyra Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# import json
+import nbformat
+import re
+
+from abc import ABC, abstractmethod
+
+from typing import Any, Type, TypeVar, List
+from traitlets import log
+
+# Setup forward reference for type hint on return from class factory method.  See
+# https://stackoverflow.com/questions/39205527/can-you-annotate-return-type-when-value-is-instance-of-cls/39205612#39205612
+F = TypeVar('F', bound='FileParser')
+
+
+class FileParser(ABC):
+
+    @classmethod
+    def get_instance(cls: Type[F], **kwargs: Any) -> F:
+        """Creates an appropriate subclass instance based on the extension of the filepath"""
+        filepath = kwargs['filepath']
+        if '.ipynb' in filepath:
+            return NotebookFileParser()
+        elif '.py' in filepath:
+            return PythonFileParser()
+        elif '.r' in filepath:
+            return RFileParser()
```

Comment:
As noted in a previous comment, let's go ahead and feed `filepath` to the constructors and remove it from `get_resources()`.

Reply:
(Sorry for being so redundant sounding here.  I still have yet to figure out how/where review comments appear.)